File: packages/typescript/src/spec.ts
export interface EnvironmentSpecInput {
  adapterId?: string;
  envOverrides?: Record<string, string>;
  seedFiles?: Record<string, string>;
  fileModes?: Record<string, number>;
  credentials?: Record<string, string>;
  prefix?: string;
}

export class EnvironmentSpec {
  readonly adapterId: string;
  readonly envOverrides: Record<string, string>;
  readonly seedFiles: Record<string, string>;
  readonly fileModes: Record<string, number>;
  readonly credentials: Record<string, string>;
  readonly prefix: string;

  constructor(input: EnvironmentSpecInput = {}) {
    this.adapterId = input.adapterId ?? "generic";
    this.envOverrides = { ...(input.envOverrides ?? {}) };
    this.seedFiles = { ...(input.seedFiles ?? {}) };
    this.fileModes = { ...(input.fileModes ?? {}) };
    this.credentials = { ...(input.credentials ?? {}) };
    this.prefix = input.prefix ?? "agent-venv-";
  }

  static fromWire(payload: Record<string, unknown> | undefined): EnvironmentSpec {
    const p = payload ?? {};
    return new EnvironmentSpec({
      adapterId: typeof p["adapter_id"] === "string" ? (p["adapter_id"] as string) : "generic",
      envOverrides: (p["env_overrides"] as Record<string, string>) ?? {},
      seedFiles: (p["seed_files"] as Record<string, string>) ?? {},
      fileModes: (p["file_modes"] as Record<string, number>) ?? {},
      credentials: (p["credentials"] as Record<string, string>) ?? {},
      prefix: typeof p["prefix"] === "string" ? (p["prefix"] as string) : "agent-venv-",
    });
  }
}

File: packages/python/src/agent_venv/spec.py
"""EnvironmentSpec: declarative description of a profile to materialize."""

from __future__ import annotations

import dataclasses
from typing import Any


@dataclasses.dataclass(slots=True)
class EnvironmentSpec:
    """Inputs to ``Environment.ephemeral`` / ``Environment.create_or_attach``.

    See spec/environment-spec.schema.json for the language-neutral contract.
    """

    adapter_id: str = "generic"
    env_overrides: dict[str, str] = dataclasses.field(default_factory=dict)
    seed_files: dict[str, str] = dataclasses.field(default_factory=dict)
    file_modes: dict[str, int] = dataclasses.field(default_factory=dict)
    credentials: dict[str, str] = dataclasses.field(default_factory=dict)
    prefix: str = "agent-venv-"

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> EnvironmentSpec:
        return cls(
            adapter_id=str(data.get("adapter_id", "generic")),
            env_overrides=dict(data.get("env_overrides") or {}),
            seed_files=dict(data.get("seed_files") or {}),
            file_modes={k: int(v) for k, v in (data.get("file_modes") or {}).items()},
            credentials=dict(data.get("credentials") or {}),
            prefix=str(data.get("prefix", "agent-venv-")),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "adapter_id": self.adapter_id,
            "env_overrides": dict(self.env_overrides),
            "seed_files": dict(self.seed_files),
            "file_modes": dict(self.file_modes),
            "credentials": dict(self.credentials),
            "prefix": self.prefix,
        }
